src/cluster_import.c

```diff
@@ -0,0 +1,1915 @@
+        } else {
+            serverLog(LL_WARNING,
+                      "Failed to snapshot slots (%s) to target %.40s",
+                      link->slot_ranges_str,
+                      link->nodename);
+            finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED, "Failed to perform snapshot");
+        }
+    }
+}
+
+int clusterIsSlotExportReadyForReplData(client *c) {
+    if (!c->flag.slot_export_target || !c->slot_migration_link) {
+        return 1;
+    }
+    slotMigrationLink *link = (slotMigrationLink *)c->slot_migration_link;
+    return link->state != SLOT_EXPORT_SNAPSHOTTING &&
+           link->state != SLOT_EXPORT_WAITING_TO_SNAPSHOT;
+}
+
+void clusterFeedSlotExportLinks(int dbid, robj **argv, int argc) {
+    UNUSED(dbid);
+    int i, error_code;
+    int slot = -1;
+    listIter li;
+    listNode *ln;
+
+    /* This function may be called after the command is executed.
+     * At this time, the arg in argv may be rewritten and the encoding
+     * may be an INT. In this case, we need to decode it into a string
+     * object because in getKeysFromCommand all the args are a string. */
+    robj **new_argv = NULL;
+    for (int i = 0; i < argc; i++) {
+        if (!sdsEncodedObject(argv[i])) {
+            new_argv = zmalloc(sizeof(robj *) * (argc));
+            break;
+        }
+    }
+    if (new_argv) {
+        for (int i = 0; i < argc; i++) new_argv[i] = getDecodedObject(argv[i]);
+        argv = new_argv;
+    }
+
+    /* Check the slot this command belongs to. Note that it is not a guarantee
+     * that the slot of the replicated command is the same as the slot of the
+     * executed command, for example in the case of module VM_Replicate APIs.
+     * Because of this case, we need to redo the slot lookup completely at this
+     * time. */
+    struct serverCommand *cmd = lookupCommand(argv, argc);
+    getNodeByQuery(server.current_client, cmd, argv, argc, &slot, &error_code);
+    if (error_code != CLUSTER_REDIR_NONE || slot == -1) {
+        /* A couple cases where this could happen:
+         *    - The replicated command is a command without a slot.
+         *    - The replicated command is written by VM_Replicate module APIs
+         *      and is a cross-slot command, or a slot that is not owned by
+         *      this node.
+         *
+         * In any case, our best solution is to not replicate this to the
+         * target node. */
+        return;
+    }
+
+    listRewind(server.cluster->slot_migration_links, &li);
+    while ((ln = listNext(&li))) {
+        slotMigrationLink *link = (slotMigrationLink *)ln->value;
+        if (link->type != SLOT_MIGRATION_EXPORT) continue;
+        if (!link->client) continue;
+        if (!isSlotMigrationLinkInProgress(link)) continue;
+        if (!isSlotInSlotRanges(slot, link->slot_ranges)) continue;
+
+        addReplyArrayLen(link->client, argc);
+        for (i = 0; i < argc; i++) {
+            addReplyBulk(link->client, argv[i]);
+        }
```

Comment:
I thought we went to great lengths to optimize this code path in replication.c, since all of these functions have overhead. Not needed here, but something to consider optimizing.

Reply:
Yeah as I mentioned above - this function probably needs an optimization pass. It is definitely usable now, but getting it to be as good as replication would be the goal